crypto/ec/ec_lib.c

```diff
@@ -282,6 +343,34 @@ int EC_GROUP_set_generator(EC_GROUP *group, const EC_POINT *generator,
         return 0;
     }
 
+    /* require group->field >= 1 */
+    if (group->field == NULL || BN_is_zero(group->field)
+        || BN_is_negative(group->field)) {
+        ECerr(EC_F_EC_GROUP_SET_GENERATOR, EC_R_INVALID_FIELD);
+        return 0;
+    }
+
+    /*-
+     * - require order >= 1
+     * - enforce upper bound due to Hasse thm: order can be no more than one bit
+     *   longer than field cardinality
+     */
+    if (order == NULL || BN_is_zero(order) || BN_is_negative(order)
+        || BN_num_bits(order) > BN_num_bits(group->field) + 1) {
+        ECerr(EC_F_EC_GROUP_SET_GENERATOR, EC_R_INVALID_GROUP_ORDER);
+        return 0;
+    }
```

Comment:
We discussed this f2f with @bbbrumley and I support this change: we are basically repeating here all the checks that the library is already doing during ASN.1 parsing for the order field before calling this function.

The previous behavior of allowing to programmatically set the `order` to zero by passing `NULL` is just silly: we can't do anything with an `EC_GROUP` object with `order == 0` anyway, and it was possible to create it only programmatically not parsing a file.

This is something an other reviewer might want to evaluate closely, but I totally support this change also for backporting to the other active branches.

Reply:
Hmm. That would be a technical compatibility break. It's not unreasonable to expect that some applications may actually set these values in multiple separate calls to this function, i.e. just because order is NULL on one call, it doesn't mean it will still be NULL by the time the group is actually used.

Reply:
> just because order is NULL on one call, it doesn't mean it will still be NULL by the time the group is actually used

I understand. But how else would `order` be set?

Reply:
So consider this code:

    EC_GROUP_set_generator(group, gen, NULL, cofactor);
    ...sometime later....
    EC_GROUP_set_generator(group, gen, order, NULL);

I'm not sure if it makes any sense to code things that way, but theoretically this was possible before and now the first call would fail.

When I wrote the above comment I was actually thinking that the generator could be done that way too, which then makes more sense, i.e. 

    EC_GROUP_set_generator(group, gen, NULL, NULL);
    ...sometime later....
    EC_GROUP_set_generator(group, NULL, order, cofactor);

But I now realise that the generator must always be present.

Reply:
Ah! No! I'm talking rubbish because of this code:

````
    if (order != NULL) {
        if (!BN_copy(group->order, order))
            return 0;
    } else {
        BN_zero(group->order);
    }

    /* The cofactor is an optional field, so it should be able to be NULL. */
    if (cofactor != NULL) {
        if (!BN_copy(group->cofactor, cofactor))
            return 0;
    } else {
        BN_zero(group->cofactor);
    }
````
i.e. if one or the other of order/cofactor is not present then it is reset to zero

So, I withdraw my objection to this.

Reply:
Great :+1: 

I will submit a separate PR for `EC_GROUP_set_generator` documentation:

> EC_GROUP_set_generator sets curve parameters that must be agreed by all participants using the curve. These parameters include the generator, the order and the cofactor. The generator is a well defined point on the curve chosen for cryptographic operations. Integers used for point multiplications will be between 0 and n-1 where n is the order. The order multiplied by the cofactor gives the number of points on the curve.

... since without digging in code it's not clear what's allowed. It's like 18yo code so it's not exactly easy to speculate ;)